### PR TITLE
Use input type in coder inference for MapElements and FlatMapElements

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

It seems that PR #756 triggered some very strange serialization issues with regards to lambda (https://builds.apache.org/job/beam_PreCommit_MavenVerify/2610/console). To isolate them, I went through the conflicts and peeled this commit off the top. Description follows:

Previously, the input `TypeDescriptor` was unknown, so we would fail to infer a coder for things like `MapElements.of(SimpleFunction<T, T>)` even if the input `PCollection` provided a coder for `T`.

Now, the input type is plumbed appropriately and the coder is inferred.

This required internal changes to explicitly support good display data (tests fail with a naive refactor). While doing this, I just added display data to `SimpleFunction` by analogy with `DoFn`.

### CODE REVIEW

**Comment by kennknowles** [comment]
A bit of flavor on the motivation by this change, which might otherwise look bigger than it needs to be:

 - `ParDo.apply` always clones its argument immediately via serialization to support the (very questionable IMO) use case of a user who wants to repeatedly mutate their `DoFn` while re-using it, and have each of the re-uses keep the value at the time of `apply`. To me, this contradicts the general semantics of Java...
 - But anyhow in that case, transient fields end up `null` so I had to make the types non-transient.
 - But even though `TypeDescriptor` is serializable, `TypeVariable` is not (this is why they were transient in the first place). And this fix is _specifically_ for cases where both the input type and output type contain type variables.
 - So I "swapped the polarity" on whether the underlying transform stashes a `SimpleFunction` or a `SerializableFunction` plus type descriptors.
 - That broke display data for a minute, since it looked directly at the classes, and in the process of fixing that, I just added display data to the functions. The fix didn't actually require this latter bit.

So this is my best approach to getting a good `TypeDescriptor` for both input and output without every having to serialize them.

**Comment by kennknowles** [comment]
R: @bjchambers AND @swegner

**Comment by swegner** [comment]
Looks like these same change appears in #756; I've added comments there.

**Comment by kennknowles** [comment]
Thanks to your comments in #756, I've fixed this one up. They are now disjoint, but this one will have to merge first.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java [hunk 1]
```diff
@@ -103,6 +106,19 @@ public void testFlatMapFnOutputTypeDescriptor() throws Exception {
   }
 
   @Test
+  public void testSimpleFunctionClassDisplayData() {
+    SimpleFunction<Integer, List<Integer>> simpleFn = new SimpleFunction<Integer, List<Integer>>() {
+      @Override
+      public List<Integer> apply(Integer input) {
+        return Collections.emptyList();
+      }
+    };
```
Comment (swegner) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java]: Can you update these tests to also verify display data items from the SimpleFunction are registered? i.e., add a `populateDisplayData()` override to add some display data and make sure it gets wired through.

**Comment by swegner** [comment]
Aside from comment about additional testing, LGTM

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java (hunk 1 above)
Comment (swegner) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java]: Ah, nevermind I see you moved that the a separate PR.

**Comment by kennknowles** [comment]
Pinging @bjchambers for binding LGTM

**Comment by kennknowles** [comment]
Travis failure is timeout.

**Comment by kennknowles** [comment]
Rebased to kick Travis.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java [hunk 2]
```diff
@@ -103,6 +106,19 @@ public void testFlatMapFnOutputTypeDescriptor() throws Exception {
   }
 
   @Test
+  public void testSimpleFunctionClassDisplayData() {
```
Comment (bjchambers) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java]: It seems like there should be tests for the "propagate input coder to output" logic in at least a few diffferent cases T -> T, T -> Iterable<T>, etc.

Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java]: Done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java [hunk 3]
```diff
@@ -102,6 +105,51 @@ public void testFlatMapFnOutputTypeDescriptor() throws Exception {
     pipeline.run();
   }
 
+  /**
+   * A {@link SimpleFunction} to test that the coder registry can propagate coders
+   * that are bound to type variables.
+   */
+  private static class PolymorphicSimpleFunction<T> extends SimpleFunction<T, Iterable<T>> {
```
Comment (bjchambers) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/FlatMapElementsTest.java]: We could also play with a few other simple functions:

SimpleFunction<T, T>
SimpleFunction<Iterable<T>, T> <- wouldn't expect this to work?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java [hunk 4]
```diff
@@ -53,6 +53,29 @@
   public transient ExpectedException thrown = ExpectedException.none();
 
   /**
+   * A {@link SimpleFunction} to test that the coder registry can propagate coders
+   * that are bound to type variables.
+   */
+  private static class PolymorphicSimpleFunction<T> extends SimpleFunction<T, T> {
```
Comment (bjchambers) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java]: Maybe a little odd that here `PolymorphicSimpleFunction = SimpleFunction<T, T>` but above `PolymorphicSimpleFunction = SimpleFunction<T, Iterable<T>>`. Looks fine either way.

**Comment by bjchambers** [comment]
Jenkins is red.

**Comment by kennknowles** [comment]
Jenkins rightly took issue with a poorly place linebreak. Fixed.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java (hunk 4 above)
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java]: Changed all the names, including s/polymorphic/generic to be more Java-ish.

**Comment by kennknowles** [comment]
Rebased onto the new `DoFn`. PTAL.
